12-offer-encoding.md

```diff
@@ -0,0 +1,769 @@
+the writer doesn't have to check all the fields are duplicates before
+simply returning a previous invoice.  Note that such caching is optional,
+and should be carefully limited when e.g. currency conversion is involved,
+or if the invoice has expired.
+
+The invoice duplicates fields rather than committing to the previous offer or
+invoice_request.  This flattened format simplifies storage at some space cost, as
+the payer need only remember the invoice for any refunds or proof.
+
+The reader of the invoice cannot trust the invoice correctly reflects the
+offer and invoice_request fields, hence the requirements to check that they
+are correct.
+
+Note that the recipient of the invoice can determine the expected
+amount from either the offer it received, or the invoice_request it
+sent, so often already has authorization for the expected amount.
+
+It's natural to set the `relative_expiry` of an invoice for a
+recurring offer to the end of the payment window for the period, but
+if that is a long time and the offer was in another currency, it's
+common to cap this at some maximum duration.  For example, omitting it
+implies the default of 7200 seconds, which is generally a sufficient
+time for payment.
+
+It's often useful to provide capacity hints, particularly where more
+than one blinded path is included, for payers to use multi-part
+payments.
+
+The invoice issuer is allowed to ignore `payer_note` (it has an odd
+number, so is optional), but if it does not, it must copy it exactly
+as the invoice_request specified.
+
+# Invoice Errors
+
+Informative errors can be returned in an onion message `invoice_error`
+field (via the onion `reply_path`) for either `invoice_request` or
+`invoice`.
+
+## TLV Fields for `invoice_error`
+
+1. `tlv_stream`: `invoice_error`
+2. types:
+    1. type: 1 (`erroneous_field`)
+    2. data:
+        * [`tu64`:`tlv_fieldnum`]
+    1. type: 3 (`suggested_value`)
+    2. data:
+        * [`...*byte`:`value`]
+    1. type: 5 (`error`)
+    2. data:
+        * [`...*utf8`:`msg`]
+
+## Requirements
+
+A writer of an invoice_error:
+  - MUST set `error` to an explanatory string.
+  - MAY set `erroneous_field` to a specific field number in the
+    `invoice` or `invoice_request` which had a problem.
+  - if it sets `erroneous_field`:
+    - MAY set `suggested_value`.
+    - if it sets `suggested_value`:
+      - MUST set `suggested_value` to a valid field for that `tlv_fieldnum`.
+  - otherwise:
+    - MUST NOT set `suggested_value`.
+
+A reader of an invoice_error:
+   FIXME!
+
+## Rationale
+
+Usually an error message is sufficient for diagnostics, however there
+is at least one case where it should be programatically parsable.  A
+recurring offer which sets `send_invoice` can also specify a currency,
+which opens the possibility for a disagreement on exchange rate.  In
+this case, the `suggested_value` reflects its expected value, and the
+sender can send a new invoice.
```

Comment:
Could this also be used when trying to pay an invoice that expires in-flight? I am thinking of the case where there is an exchange rate lock-in for e.g. 15 minutes, but although the invoice payment is initiated within those 15 minutes, it actually reaches the receiving node after it expired. It could therefore return an error with the new suggested value for amount using an updated exchange rate?

Reply:
Definitely.